### PR TITLE
Implement `RetAbi` and `BoxRet`

### LINKED ISSUE #1484: Remove TableIterator-to-SetOfIterator remapping codegen?
I think the way #1091 was solved leaves some lingering questions.

The error here is emitted from our code:

https://github.com/pgcentralfoundation/pgrx/blob/4fa87bd57d5e673e280c987a75cd8129ff349f43/pgrx/src/srf.rs#L96-L100

But we apparently "need" to convert the TableIterator into a SetOfIterator: https://github.com/pgcentralfoundation/pgrx/blob/5a45b3ae7ac3fe120181f4c50df689c1fb067c78/pgrx-sql-entity-graph/src/pg_extern/mod.rs#L557-L567

I find this to be a somewhat unclear statement, as it's not clear if the obligation is coming from Postgres or from our own code. It is strongly implied to be "from Postgres":

https://github.com/pgcentralfoundation/pgrx/blob/5a45b3ae7ac3fe120181f4c50df689c1fb067c78/pgrx-sql-entity-graph/src/pg_extern/mod.rs#L547-L549

However, what precisely seems to be happening is that we're the ones making the assumption that the return type must be a composite type if we're using a TableIterator. If this understanding is correct, then I think it should be written down as such. It also opens up an alternative solution that can move handling the one-item-tuple case out of the generated code, by using the fact this code is a trait impl and thus we can make generic impls do the lifting. And solving fewer problems during codegen is always preferable.

cc @eeeebbbbrrrr

### PR DESCRIPTION
**Pull request by workingjubilee**
This completely reworks the way we handle returning types from functions, so that we no longer have to rely on a macro expansion behavior that has to already know the types at expansion time (and thus has to parse them somehow, which it cannot realistically do because type aliases exist). Instead, we simply expand into code that asks the types themselves to modify the FunctionCallInfo appropriately and then return a raw Datum to Postgres.

This breaks support for certain returns because it was difficult to do this and also support arbitrary nesting, because Postgres does not support arbitrary nesting. For instance, you can no longer return:
- `SetOfIterator<'a, Result<T, E>>`
- `TableIterator<'a, (Result<T, E>, Result<U, D>)>`
- `Option<TableIterator<'a, Tuple>>`

It's expected that this will improve in the near-ish future.

This also breaks returning values from `#[pg_extern]` functions that were relying on `IntoDatum` implementations being enough. It is not expected this will improve, for the reasons described on the documentation of the new traits, which can be summarized as "`IntoDatum` should never have been used for that bound". This change blocks off several latent correctness problems from affecting pgrx going forward.

Fixes https://github.com/pgcentralfoundation/pgrx/issues/1484

### CODE REVIEW

File: pgrx/src/callconv.rs
Comment (workingjubilee) [inline pgrx/src/callconv.rs]: This is what I meant about the awkwardness of forcing the scalar type cases into a framework that can support the iterator and composite type cases.

File: pgrx/src/callconv.rs
Comment (workingjubilee) [inline pgrx/src/callconv.rs]: Most of these being should-never-happen cases, effectively.

File: pgrx/src/iter.rs
Comment (workingjubilee) [inline pgrx/src/iter.rs]: Probably not even correct, here. We don't have enough tests for this edge-case.

File: pgrx/src/iter.rs
Comment (workingjubilee) [inline pgrx/src/iter.rs]: This is the most awkward thing that actually makes me think "dealbreaker, maybe?" until I restructure it: the tuple types shouldn't have to have magical knowledge of the TupleDesc that was set up for them by TableIterator.

File: pgrx/src/iter.rs
Comment (workingjubilee) [inline pgrx/src/iter.rs]: so many cases of this...

File: pgrx/src/callconv.rs
Comment (workingjubilee) [reply pgrx/src/callconv.rs]: I resolved this by using a blanket impl.

File: pgrx/src/iter.rs
Comment (workingjubilee) [inline pgrx/src/iter.rs]: This is... not ideal, as it traps us in a position where we might accidentally try to return something like `TableIterator<'a, (TableIterator<'b, TupleA>, TableIterator<'c, TupleB>)>`.

Comment (workingjubilee) [reply pgrx/src/iter.rs]: This would be preferable, as it closes off that route:
```suggestion
    C: RetPackage,
```

Comment (workingjubilee) [reply pgrx/src/iter.rs]: I should note that I'm pretty sure that this is currently impossible anyways even with this PR because the schema generator chokes on it, I just would rather it didn't require "the schema generator chokes on it" in order to stop us from doing something silly.

File: pgrx-tests/src/tests/srf_tests.rs
Comment (workingjubilee) [inline pgrx-tests/src/tests/srf_tests.rs:297]: So, this _should_ be `TableIterator<'_, Result<Tuple, Error>>` so that no eager evaluation has to happen. I think I can make that work with a bit more effort. But the schema generator chokes on **that**! And that is another reason why I want it out of our way.

I do not think people should be required to return `TableIterator<'a, (Result, Result, Result, Result, Result, Result)>` if they want to fallibly return a row, but they currently _are_. And I do not think this should be blocked on that.

Asking people to propagate errors to the top of each row return should be fine.

File: pgrx-examples/spi_srf/src/lib.rs
Comment (workingjubilee) [inline pgrx-examples/spi_srf/src/lib.rs:114]: An advantage of using a somewhat more sensible return type is that you don't have this sort of thing happen...

**Review by workingjubilee — COMMENTED** [review]
As the iterator conveniences are basically independent, I'll rebase this on #1703 after that lands.

File: pgrx/src/iter.rs
Comment (workingjubilee) [reply pgrx/src/iter.rs]: I intend to fix this before we land it but I believe otherwise this is ready for review, I don't want to do much more work on this without feedback.

**Comment by workingjubilee** [comment]
This PR opens a LOT of pure simplifications like 9dd8f5fd055748bf6cd3891acb85d3f731f5d782, it's just going to take forever to find them all.

File: pgrx-tests/src/tests/srf_tests.rs
Comment (eeeebbbbrrrr) [reply pgrx-tests/src/tests/srf_tests.rs:297]: > Asking people to propagate errors to the top of each row return should be fine.

I'm not sure this is a nice thing to ask.  When SPI, for example, is involved, and what you're trying to do is return a TableIterator of a bunch of individual values you somehow got from SPI, it's definitely a lot easier to just return the `Result<T>` (or `Result<Option<T>>`) in whatever tuple position.  Especially if there's quite a few values in the tuple.

This isn't a hill I'm willing to die on, but it'd be nice to support, for example, `TableIterator<(Result<S>, T, Result<U>, Result<V>, W, Result<X>, Y, Result<Z>)>`.  Then we'll panic whenever we get to unwrapping that result and the user doesn't have to worry about it.

Comment (workingjubilee) [reply pgrx-tests/src/tests/srf_tests.rs:297]: @eeeebbbbrrrr Honestly, we'll never have good UX until we expand to supporting actual `struct` arguments, really.

Comment (eeeebbbbrrrr) [reply pgrx-tests/src/tests/srf_tests.rs:297]: What does "actual `struct` arguments" mean?

File: pgrx-examples/spi_srf/src/lib.rs
Comment (workingjubilee) [inline pgrx-examples/spi_srf/src/lib.rs:99]: @eeeebbbbrrrr it's certainly not less code but in most cases it just looks like going "hello? I? have? a? question? if that's okay?"

File: pgrx-tests/src/tests/srf_tests.rs
Comment (workingjubilee) [reply pgrx-tests/src/tests/srf_tests.rs:297]: TableIterator should be able to return a struct that describes the row in question, but this requires being able to use traits like these to return fairly arbitrary generic types. It doesn't work as long as we rely on parsing.

File: pgrx-examples/spi_srf/src/lib.rs
Comment (eeeebbbbrrrr) [reply pgrx-examples/spi_srf/src/lib.rs:99]: It's not terrible, but I'm sure you agree that returning a Result from a closure is awkward at best.  In this case we `.collect::<Result<Vec<_>>>()` so it's not too terrible.  

I prefer the left side of this diff, but like I said elsewhere, I'm not willing to die on this hill.

Comment (workingjubilee) [reply pgrx-examples/spi_srf/src/lib.rs:99]: yeah, the left side is a bit nicer, but that type signature... that's what I really have it in for. so yeah, if I didn't believe it would open up better possibilities in short order, I wouldn't do it.

Comment (eeeebbbbrrrr) [reply pgrx-examples/spi_srf/src/lib.rs:99]: > that's what I really have it in for

Yeah, that's why I'm not really gonna fight on this one.  I hate typing out all that mess too.  But once the return signature has been typed out, it makes it a lot easier to just write some code and return some values.

File: pgrx/src/callconv.rs
Comment (workingjubilee) [inline pgrx/src/callconv.rs]: uh that is extremely not correct
```suggestion
        let iter = &mut *(*fcx).user_fctx.cast::<SetOfIterator<'_, T>>();
```

**Comment by workingjubilee** [comment]
I moved the implementations for the iterator types to live next to the iterators. It seemed sensical for some reason.

**Comment by eeeebbbbrrrr** [comment]
Ya know. What if we called it…

I’m kidding. This is great. Thanks, @workingjubilee

**Comment by workingjubilee** [comment]
Resolved the "tuples have psychic powers" issue by splitting out their knowledge of how to add a tupdesc into a new fn.

File: pgrx-tests/src/tests/srf_tests.rs
Comment (workingjubilee) [inline pgrx-tests/src/tests/srf_tests.rs:309]: Breaking this one annoys me much more than the TableIterator nesting case but I want to make sure that SetOfIterator has exactly as stringent requirements, currently, as `TableIterator<'a, (T,)>`, because it will make things much worse down the line if they meaningfully differ.

File: pgrx-examples/spi_srf/src/lib.rs
Comment (NotGyro) [reply pgrx-examples/spi_srf/src/lib.rs:99]: Many question marks is much-preferable to having excessively-verbose nested type signatures, yeah.

**Review by NotGyro — APPROVED** [review]
Looks good to me. This really does make code working with the returns from Rust to Postgres much cleaner and more elegant. 

I had some confusion about the SetOfIterator type, but it was cleared up in a company meeting. The single-item case behaves differently than the iterator case for legacy Postgres reasons, which is why SetOfIterator can be one item and/or an iterator.

**Review by NotGyro — APPROVED** [review]
Thanks for the documentation! This explains it neatly.

**Comment by workingjubilee** [comment]
This last commit makes the "Ret" type into an associated type and thus makes the iterators eat all the complexity of their return ABI as now only they have to know about how to drive themselves forward. `Result` does not have to think about them, and the implementation details are once more private to iterators and fully hidden behind the trait. This prevents anyone from trying to make any of the **many** bug-inducing nonsensical calls that were technically allowed at first, and probably made the code have to carry around way more panic branches.

**Comment by workingjubilee** [comment]
The last few commits were just cleanup.

**Pull request merged**
